src/freehand_selection.jl

```diff
@@ -1,61 +1,188 @@
 function init_freehand_select(ctx::ImageContext)
     c = ctx.canvas
     enabled = Signal(true)
-    dragging = Signal(false)
+    drawing = Signal(false)
+    moving = Signal(false)
+    diff = Signal(XY(NaN,NaN))
 
     dummybtn = MouseButton{UserUnit}()
     
     sigstart = map(filterwhen(enabled, dummybtn, c.mouse.buttonpress)) do btn
         if !ispolygon(value(ctx.points)) || !isinside(Point(btn.position), Point.(value(ctx.points))) # prevents conflict with init_move_polygon
-            push!(dragging, true)
+            push!(drawing, true)
             push!(ctx.shape, Rectangle()) # some identifier of type of selection
             push!(ctx.points, [])
```

Comment:
Just for fun...let's define this:
```julia
foo(a, b) = XY(a.x+b.x, a.y+b.y)
```
Now let's look a the assembly language that julia generates for `XY{Float64}`:
```julia
julia> code_native(foo, Tuple{XY{Float64}, XY{Float64}})
        .text
Filename: REPL[12]
        pushq   %rbp
        movq    %rsp, %rbp
Source line: 1
        vmovsd  (%rsi), %xmm0           # xmm0 = mem[0],zero
        vmovsd  8(%rsi), %xmm1          # xmm1 = mem[0],zero
        vaddsd  (%rdx), %xmm0, %xmm0
        vaddsd  8(%rdx), %xmm1, %xmm1
        vmovsd  %xmm0, (%rdi)
        vmovsd  %xmm1, 8(%rdi)
        movq    %rdi, %rax
        popq    %rbp
        retq
        nopw    %cs:(%rax,%rax)
```
You don't have to understand this thoroughly to see that it's much shorter than the assembly generated for the abstract type `XY`:
```julia
julia> code_native(foo, Tuple{XY, XY})
; WARNING: This code may not match what actually runs.
        .text
Filename: REPL[12]
        pushq   %rbp
        movq    %rsp, %rbp
        pushq   %r15
        pushq   %r14
        pushq   %r13
        pushq   %r12
        pushq   %rbx
        subq    $168, %rsp
        movq    %rsi, %r12
        movq    %rdi, %r14
        movq    %fs:0, %rcx
        addq    $-10888, %rcx           # imm = 0xD578
        movq    %rcx, -48(%rbp)
        leaq    -64(%rbp), %rsi
        vxorps  %ymm0, %ymm0, %ymm0
        vmovups %ymm0, -88(%rbp)
        vmovups %ymm0, -120(%rbp)
        vmovups %ymm0, -152(%rbp)
        vmovups %ymm0, -184(%rbp)
        movq    $0, -56(%rbp)
        movq    $34, -200(%rbp)
        movq    (%rcx), %rax
        movq    %rax, -192(%rbp)
        leaq    -200(%rbp), %rax
        movq    %rax, (%rcx)
Source line: 1
        movq    %r14, -64(%rbp)
        movabsq $140372008327048, %r15  # imm = 0x7FAAE7B15788
        movq    %r15, -56(%rbp)
        movabsq $jl_f_getfield, %r13
        xorl    %edi, %edi
        movl    $2, %edx
        vzeroupper
        callq   *%r13
        movq    %rax, -152(%rbp)
        movq    %r12, -80(%rbp)
        movq    %r15, -72(%rbp)
        xorl    %edi, %edi
        movl    $2, %edx
        leaq    -80(%rbp), %rsi
        callq   *%r13
        movq    %rax, -144(%rbp)
        leaq    3834048(%r15), %r13
        movq    %r13, -160(%rbp)
        movabsq $jl_apply_generic, %rax
        movl    $3, %esi
        leaq    -160(%rbp), %rdi
        callq   *%rax
        movq    %rax, -128(%rbp)
        movq    %r14, -112(%rbp)
        leaq    849520(%r15), %rbx
        movq    %rbx, -104(%rbp)
        xorl    %edi, %edi
        movl    $2, %edx
        leaq    -112(%rbp), %rsi
        movabsq $jl_f_getfield, %rax
        callq   *%rax
        movq    %rax, -176(%rbp)
        movq    %r12, -96(%rbp)
        movq    %rbx, -88(%rbp)
        xorl    %edi, %edi
        movl    $2, %edx
        leaq    -96(%rbp), %rsi
        movabsq $jl_f_getfield, %rax
        callq   *%rax
        movq    %rax, -168(%rbp)
        movq    %r13, -184(%rbp)
        movl    $3, %esi
        leaq    -184(%rbp), %rdi
        movabsq $jl_apply_generic, %rbx
        callq   *%rbx
        movq    %rax, -120(%rbp)
        addq    $-292313880, %r15       # imm = 0xEE93A4E8
        movq    %r15, -136(%rbp)
        movl    $3, %esi
        leaq    -136(%rbp), %rdi
        callq   *%rbx
        movq    -192(%rbp), %rcx
        movq    -48(%rbp), %rdx
        movq    %rcx, (%rdx)
        addq    $168, %rsp
        popq    %rbx
        popq    %r12
        popq    %r13
        popq    %r14
        popq    %r15
        popq    %rbp
        retq
```
It's even worse than it looks because those `callq` statements mean "call another function" whereas the first one is entirely self-contained.